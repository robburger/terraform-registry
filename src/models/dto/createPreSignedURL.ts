export interface CreatePreSignedURLDTO {
  namespace: string
  name: string
  version: string
  provider: string
}
