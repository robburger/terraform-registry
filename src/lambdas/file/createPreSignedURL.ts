import { APIGatewayProxyEvent, APIGatewayProxyResult } from 'aws-lambda'

import { CreatePreSignedURLDTO } from '../../models/dto'
import { AWS, Logger, Message } from '../../utils'

const logger = new Logger()

const s3 = new AWS.S3({ region: process.env.AWS_REGION, signatureVersion: 'v4' })

export const handler = async (event: APIGatewayProxyEvent): Promise<APIGatewayProxyResult> => {
  if (event.queryStringParameters?.type === 'module') {
    const params: CreatePreSignedURLDTO = JSON.parse(event.body)
    if (isValidParams(params)) {
      try {
        const ext = event.queryStringParameters?.compression || 'zip'
        const url = s3.getSignedUrl('putObject', {
          Bucket: process.env.S3_BUCKET_NAME,
          Key: `modules/${params.namespace}/${params.name}/${params.provider}/${params.version}/module.${ext}`,
          Expires: 300,
          Tagging: 'tags=true',
        })
        logger.info(`Successfully generated pre-signed URL`)
        return Message.success({
          url: url,
        })
      } catch (e) {
        logger.error(`Failed to generate pre-signed URL: ${e}`)
        return Message.error(`Failed to generate pre-signed URL: ${e}`)
      }
    } else {
      logger.error(`JSON payload invalid: ${JSON.stringify(params)}`)
      return Message.bad(
        `JSON payload invalid. Ensure you provide an object with 'namespace', 'name', 'provider' and 'version' parameters`,
      )
    }
  } else {
    logger.error(`Missing 'type' from queryStringParameters`)
    return Message.bad(
      `You need to specify a 'type' for this request. e.g. /pre-sign?type=module or /pre-sign?type=provider`,
    )
  }
}

const isValidParams = (params: CreatePreSignedURLDTO) => {
  if (
    params?.hasOwnProperty('namespace') &&
    params?.namespace.length > 0 &&
    params?.hasOwnProperty('name') &&
    params?.name.length > 0 &&
    params?.hasOwnProperty('version') &&
    params?.version.length > 0 &&
    params?.hasOwnProperty('provider') &&
    params?.provider.length > 0
  ) {
    return true
  } else {
    return false
  }
}
