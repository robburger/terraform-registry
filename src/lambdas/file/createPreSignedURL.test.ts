import { APIGatewayProxyEvent } from 'aws-lambda'
import { handler } from './createPreSignedURL'

import * as data from '../../../test/events/file/createPreSignedURL/CreatePreSignedURL.json'
import * as data_invalid_body from '../../../test/events/file/createPreSignedURL/CreatePreSignedURL_invalid_body.json'
import * as data_missing_body from '../../../test/events/file/createPreSignedURL/CreatePreSignedURL_missing_body.json'
import * as data_missing_querystring from '../../../test/events/file/createPreSignedURL/CreatePreSignedURL_missing_querystring.json'

beforeAll(() => {
  process.env['S3_BUCKET_NAME'] = 'localbucket'
})

beforeEach(() => {
  jest.useFakeTimers()
})

describe('createPreSignedURL handler', () => {
  it('should return a pre-signed URL', async () => {
    const mockEvent: APIGatewayProxyEvent = data
    const result = await handler(mockEvent)
    expect(result).toHaveProperty('body')
    expect(result).toHaveProperty('headers', { 'content-type': 'application/json' })
    expect(result).toHaveProperty('statusCode', 200)

    const body = JSON.parse(result.body)
    expect(body).toHaveProperty('url')
    expect(body.url).toMatch(
      /localbucket\.s3\..*amazonaws\.com\/modules\/test\/test\/test\/test\/module\.zip\?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=.*%2F\d{8}%2F\w{2}-\w*-\d%2Fs3%2Faws4_request&X-Amz-Date=\d{8}T\d{6}Z&X-Amz-Expires=\d*&X-Amz-Signature=.{64}&X-Amz-SignedHeaders=host%3Bx-amz-tagging&x-amz-tagging=tags%3Dtrue/,
    )
  })

  it('should return an error when invalid body data', async () => {
    const mockEvent: APIGatewayProxyEvent = data_invalid_body
    const result = await handler(mockEvent)
    expect(result).toMatchSnapshot()
  })

  it('should return an error when missing body data', async () => {
    const mockEvent: APIGatewayProxyEvent = data_missing_body
    const result = await handler(mockEvent)
    expect(result).toMatchSnapshot()
  })

  it('should return an error when missing/invalid "type" querystring', async () => {
    const mockEvent: APIGatewayProxyEvent = data_missing_querystring
    const result = await handler(mockEvent)
    expect(result).toMatchSnapshot()
  })
})
