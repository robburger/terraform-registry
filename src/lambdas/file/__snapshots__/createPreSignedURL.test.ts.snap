// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createPreSignedURL handler should return an error when invalid body data 1`] = `
Object {
  "body": "{\\"error\\":\\"JSON payload invalid. Ensure you provide an object with 'namespace', 'name', 'provider' and 'version' parameters\\"}",
  "headers": Object {
    "content-type": "application/json",
  },
  "statusCode": 400,
}
`;

exports[`createPreSignedURL handler should return an error when missing body data 1`] = `
Object {
  "body": "{\\"error\\":\\"JSON payload invalid. Ensure you provide an object with 'namespace', 'name', 'provider' and 'version' parameters\\"}",
  "headers": Object {
    "content-type": "application/json",
  },
  "statusCode": 400,
}
`;

exports[`createPreSignedURL handler should return an error when missing/invalid "type" querystring 1`] = `
Object {
  "body": "{\\"error\\":\\"You need to specify a 'type' for this request. e.g. /pre-sign?type=module or /pre-sign?type=provider\\"}",
  "headers": Object {
    "content-type": "application/json",
  },
  "statusCode": 400,
}
`;
